akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,325 @@
+ *
+ */
+@InternalApi private[akka] object EventsourcedRunning {
+
+  final case class EventsourcedState[State](
+    seqNr: Long,
+    state: State
+  ) {
+
+    def nextSequenceNr(): EventsourcedState[State] =
+      copy(seqNr = seqNr + 1)
+
+    def updateLastSequenceNr(persistent: PersistentRepr): EventsourcedState[State] =
+      if (persistent.sequenceNr > seqNr) copy(seqNr = persistent.sequenceNr) else this
+
+    def applyEvent[C, E](setup: EventsourcedSetup[C, E, State], event: E): EventsourcedState[State] = {
+      val updated = setup.eventHandler(state, event)
+      copy(state = updated)
+    }
+  }
+
+  def apply[C, E, S](setup: EventsourcedSetup[C, E, S], state: EventsourcedState[S]): Behavior[InternalProtocol] =
+    new EventsourcedRunning(setup).handlingCommands(state)
+}
+
+// ===============================================
+
+/** INTERNAL API */
+@InternalApi private[akka] class EventsourcedRunning[C, E, S](
+  override val setup: EventsourcedSetup[C, E, S])
+  extends EventsourcedJournalInteractions[C, E, S] with EventsourcedStashManagement[C, E, S] {
+  import EventsourcedRunning.EventsourcedState
+
+  private def log = setup.log
+  private def commandContext = setup.commandContext
+
+  def handlingCommands(state: EventsourcedState[S]): Behavior[InternalProtocol] = {
+
+    def onCommand(state: EventsourcedState[S], cmd: C): Behavior[InternalProtocol] = {
+      val effect = setup.commandHandler(commandContext, state.state, cmd)
+      applyEffects(cmd, state, effect.asInstanceOf[EffectImpl[E, S]]) // TODO can we avoid the cast?
+    }
+
+    @tailrec def applyEffects(
+      msg:         Any,
+      state:       EventsourcedState[S],
+      effect:      EffectImpl[E, S],
+      sideEffects: immutable.Seq[ChainableEffect[_, S]] = Nil
+    ): Behavior[InternalProtocol] = {
+      if (log.isDebugEnabled)
+        log.debug(s"Handled command [{}], resulting effect: [{}], side effects: [{}]", msg.getClass.getName, effect, sideEffects.size)
+
+      effect match {
+        case CompositeEffect(eff, currentSideEffects) ⇒
+          // unwrap and accumulate effects
+          applyEffects(msg, state, eff, currentSideEffects ++ sideEffects)
+
+        case Persist(event) ⇒
+          // apply the event before persist so that validation exception is handled before persisting
+          // the invalid event, in case such validation is implemented in the event handler.
+          // also, ensure that there is an event handler for each single event
+          val newState = state.applyEvent(setup, event)
+          val eventToPersist = tagEvent(event)
+
+          val newState2 = internalPersist(newState, eventToPersist)
+
+          val handler: Any ⇒ Unit = { _ ⇒
+            if (setup.snapshotWhen(newState2.state, event, newState2.seqNr))
+              internalSaveSnapshot(newState2)
+          }
+          val pendingInvocations = StashingHandlerInvocation(event, handler) :: Nil
```

Comment:
Don't use that untyped `StashingHandlerInvocation` class here.
I think this can be simplified. We only support one `Persist` or `PersistAll` effect, right?
So we only need one callback, not a list.
And we don't care about the event.
Meaning that `pendingInvocations` can be `afterPersist: () => Unit`

Reply:
cont a few lines below...